src/mono/mono/metadata/image.c

```diff
@@ -1846,34 +1846,74 @@ mono_image_loaded_by_guid (const char *guid)
 	return mono_image_loaded_by_guid_internal (guid, FALSE);
 }
 
+static const char *
+get_image_culture (MonoImage *image)
+{
+	MonoTableInfo *t = &image->tables [MONO_TABLE_ASSEMBLY];
+	if (!t->rows)
+		return NULL;
+
+	guint32 cols [MONO_ASSEMBLY_SIZE];
+	mono_metadata_decode_row (t, 0, cols, MONO_ASSEMBLY_SIZE);
+	return mono_metadata_string_heap (image, cols [MONO_ASSEMBLY_CULTURE]);
+}
+
+char *
+mono_image_get_name_with_culture_if_needed (MonoImage *image)
+{
+	if (!g_str_has_prefix (image->name, "data-") &&
+		!g_path_is_absolute (image->name))
+	{
+		const char *culture = get_image_culture (image);
+
+		if (culture && culture [0] != 0)
+			return g_strdup_printf ("%s/%s", culture, image->name);
+	}
+
+	return NULL;
+}
+
 static MonoImage *
 register_image (MonoLoadedImages *li, MonoImage *image, gboolean *problematic)
 {
 	MonoImage *image2;
+	char *name = image->name;
+#ifdef ENABLE_NETCORE
+	char *name_with_culture = mono_image_get_name_with_culture_if_needed (image);
```

Comment:
I would add some sort of note about about this since it's basically a hack to get around the fact we only register images based on the filename. It's quite possible someone is confused by this down the line.

I'd probably also add a note in `mono_image_open_a_lot_parameterized` about the function's lack of awareness of this hack. If you load a satellite image by name for one culture with it and then attempt to load a second culture after, you should get the wrong image. This is not a regression, and I believe is how this has worked for a long time now, but it's the one place we *didn't* fix this.